tools/ports/vorbis.py

```diff
@@ -16,23 +16,36 @@ def get(ports, settings, shared):
     return []
 
   ports.fetch_project('vorbis', 'https://github.com/emscripten-ports/vorbis/archive/' + TAG + '.zip', 'Vorbis-' + TAG, sha512hash=HASH)
-  libname = ports.get_lib_name('libvorbis')
-
-  def create():
-    logging.info('building port: vorbis')
-
-    source_path = os.path.join(ports.get_dir(), 'vorbis', 'Vorbis-' + TAG)
-    dest_path = os.path.join(shared.Cache.get_path('ports-builds'), 'vorbis')
-
-    shutil.rmtree(dest_path, ignore_errors=True)
-    shutil.copytree(source_path, dest_path)
-
-    final = os.path.join(dest_path, libname)
-    ports.build_port(os.path.join(dest_path, 'lib'), final, [os.path.join(dest_path, 'include')],
-                     ['-s', 'USE_OGG=1'], ['psytune', 'barkmel', 'tone', 'misc'])
-    return final
-
-  return [shared.Cache.get(libname, create)]
+  lib_name = ports.get_lib_name('libvorbis')
+  lib_name_file = ports.get_lib_name('libvorbisfile')
+  lib_name_enc = ports.get_lib_name('libvorbisenc')
+
+  def create(library):
+    def internal_create():
+      logging.info('building port: vorbis')
+
+      source_path = os.path.join(ports.get_dir(), 'vorbis', 'Vorbis-' + TAG)
+      dest_path = os.path.join(shared.Cache.get_path('ports-builds'), 'vorbis')
+
+      if not create.recreated_tree:
+        logging.info('recreating tree %s %s %s' % (library, source_path, dest_path))
+        shutil.rmtree(dest_path, ignore_errors=True)
+        shutil.copytree(source_path, dest_path)
+        create.recreated_tree = True
+
+      final = os.path.join(dest_path, library)
+      excluded_files = ['psytune', 'barkmel', 'tone', 'misc']
+      if library == lib_name_file or library == lib_name:
+        excluded_files.append('vorbisenc')
+      if library == lib_name_enc or library == lib_name:
+        excluded_files.append('vorbisfile')
```

Comment:
I think the original code include the `enc` and the `file` stuff in once big library.  

If you want to split those out then don't you need to exlude both the `enc` and the `file` sources from the core library?  and also exclude the core library files from both the `enc` and the `file` libraries. 

I'm not sure if switching to some kind of whitelist might help?

Reply:
I am unsure what you mean here.

When compiling vorbis linux usage, the following libraries are compiled from the following files:
* vorbis.a: mdct.c smallft.c block.c envelope.c window.c lsp.c \
                        lpc.c analysis.c synthesis.c psy.c info.c \
                        floor1.c floor0.c\
                        res0.c mapping0.c registry.c codebook.c sharedbook.c\
                        lookup.c bitrate.c\
                        envelope.h lpc.h lsp.h codebook.h misc.h psy.h\
                        masking.h os.h mdct.h smallft.h highlevel.h\
                        registry.h scales.h window.h lookup.h lookup_data.h\
                        codec_internal.h backends.h bitrate.h
* vorbisenc.a: links vorbis.a and adds vorbisenc.c.
* vorbisfile.a: links vorbis.a and adds vorbisfile.c

Compiling all the source files used in vorbis again for vorbisenc/vorbisfile has the same end-result as linking vorbis.

Reply:
Normally each library contains a unique set of files without overlap.  Indeed this seems to be how vorbis works, at least on my desktop system:

```
$ ar t  /usr/lib/x86_64-linux-gnu/libvorbis.a
mdct.o
smallft.o
block.o
envelope.o
window.o
lsp.o
lpc.o
analysis.o
synthesis.o
psy.o
info.o
floor1.o
floor0.o
res0.o
mapping0.o
registry.o
codebook.o
sharedbook.o
lookup.o
bitrate.o
```

```
 $ ar t  /usr/lib/x86_64-linux-gnu/libvorbisfile.a
vorbisfile.o
```

```
$ ar t  /usr/lib/x86_64-linux-gnu/libvorbisenc.a
vorbisenc.o
```

The good news is that those second two libraries only have a single file in them, so it should be trivial to build them without the need for whitelists/blacklists.

Why this matters for fastcomp:  Normally duplicating object between libraries is fine due to the way .a files are linked, but the the ports system is setup to create giant object files rather than archive files when building with fastcomp.     Basically it builds libvorbis.bc not libvorbis.a.

See: https://github.com/emscripten-core/emscripten/blob/0d5dc73a07209e25ad4fe094eccf81c83b823f63/tools/system_libs.py#L86

Reply:
So it seems. I must've made an oversight somewhere. I'll correct the PR.